src/main/java/org/interledger/Condition.java

```diff
@@ -5,69 +5,103 @@
 
 /**
  * The execution condition attached to all transfers in an Interledger payment.
- * 
+ *
  * <p>Interledger relies on conditional payments where each transfer that is part of a payment is
  * conditional upon the presentation of a fulfillment.
- * 
+ *
  * <p>The standard for conditions is to use the SHA-256 hash of a pre-image. The pre-image is
  * therefor the fulfillment of the condition.
- * 
+ *
  * @see Fulfillment
  */
-public class Condition {
+public interface Condition {
 
-  private final byte[] hash;
+  /**
+   * Get the default builder.
+   *
+   * @return a {@link Builder} instance.
+   */
+  static Builder builder() {
+    return new Builder();
+  }
 
   /**
-   * Create a new {@code Condition} from a SHA-256 hash.
-   * 
-   * @param hash The SHA-256 hash of a pre-image that is used as the fulfillment of this condition
+   * Build a new Condition using the provided hash
+   *
+   * @param hash A SHA-256 hash representing a Condition.
+   *
+   * @return a {@link Condition} instance.
    */
-  public Condition(byte[] hash) {
+  static Condition of(byte[] hash) {
+    return new Builder().hash(hash)
+        .build();
+  }
 
-    Objects.requireNonNull(hash, "Hash must not be null!");
-    if (hash.length != 32) {
-      throw new IllegalArgumentException("Hash must be 32 bytes.");
-    }
+  /**
+   * Get the SHA-256 hash of this condition.
+   *
+   * @return a {@code byte[]} of exactly 32 bytes
+   */
+  byte[] getHash();
 
-    this.hash = Arrays.copyOf(hash, 32);
-  }
+  class Builder {
 
-  @Override
-  public int hashCode() {
-    final int prime = 31;
-    int result = 1;
-    result = prime * result + Arrays.hashCode(hash);
-    return result;
-  }
+    private byte[] hash;
+
+    public Builder hash(byte[] hash) {
+      this.hash = Arrays.copyOf(hash, 32);
+      return this;
 
-  @Override
-  public boolean equals(Object obj) {
-    if (this == obj) {
-      return true;
-    }
-    if (obj == null) {
-      return false;
-    }
-    if (getClass() != obj.getClass()) {
-      return false;
     }
 
-    Condition other = (Condition) obj;
-    if (!Arrays.equals(hash, other.hash)) {
-      return false;
+    public Condition build() {
+      return new Impl(this);
     }
 
-    return true;
-  }
 
-  /**
-   * Get the SHA-256 hash of this condition.
-   * 
-   * @return a {@code byte[]} of exactly 32 bytes
-   */
-  public byte[] getHash() {
-    return Arrays.copyOf(this.hash, 32);
+    private static final class Impl implements Condition {
+
+      private final byte[] hash;
+
+      public Impl(Builder builder) {
+
+        Objects.requireNonNull(builder.hash, "Hash must not be null!");
+        if (builder.hash.length != 32) {
+          throw new IllegalArgumentException("Hash must be 32 bytes.");
+        }
+
+        this.hash = Arrays.copyOf(builder.hash, 32);
```

Comment:
It's possible to call build multiple times on the builder and produce, what should be, immutable copies of the same Condition so I think you need both.

But, that said I should be able to pass the builder an object, call `build()` then change the object I provided to the builder and call `build()` again to get a different outcome. 

That seems a bit "unintentional" but strictly speaking is correct. WDYT? @sappenin @earizon